### PR TITLE
[ML] Ensure auc_roc curve is monotonic

### PR DESCRIPTION
**Pull request by dimitris-athanasiou**
As we collect points for our auc roc curve from two
different percentiles aggregations, the result may
contain points with equal threshold that are not
monotonic. This is because the percentiles aggregation
is an approximation.

This commit ensures the fina auc roc curve we calculate
is monotonic by collapsing points of equal threshold into
a single point that is the average of the equal threshold
points it represents.

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/evaluation/common/AbstractAucRoc.java [hunk 1]
```diff
@@ -74,17 +74,76 @@ public String getName() {
         assert tpPercentiles.length == fpPercentiles.length;
         assert tpPercentiles.length == 99;
 
-        List<AucRocPoint> aucRocCurve = new ArrayList<>();
-        aucRocCurve.add(new AucRocPoint(0.0, 0.0, 1.0));
-        aucRocCurve.add(new AucRocPoint(1.0, 1.0, 0.0));
+        List<AucRocPoint> points = new ArrayList<>();
```
Comment (benwtrent) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/evaluation/common/AbstractAucRoc.java:77]: #small thing

I think you would pre-size this array list to be `new ArrayList<>(tpPercentiles.length + fpPercentiles.length)`

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/evaluation/common/AbstractAucRoc.java [hunk 2]
```diff
@@ -74,17 +74,76 @@ public String getName() {
...
+     * Visible for testing
+     *
+     * Expects a sorted list of {@link AucRocPoint} points.
+     * Collapses points with equal threshold by replacing them
+     * with a single point that is the average.
+     *
+     * @param points A sorted list of {@link AucRocPoint} points
+     * @return a new list of points where equal threshold points have been collapsed into their average
+     */
+    static List<AucRocPoint> collapseEqualThresholdPoints(List<AucRocPoint> points) {
```
Comment (benwtrent) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/evaluation/common/AbstractAucRoc.java:107]: I am not 100% sure this is enough. The issue is that we return points from different thresholds with different tf, fp values that are not monotonically increasing. 

Does this really fix that? It seems to only merge points within the same thresholds (maybe I am misunderstanding the concept of thresholds). 

A true rolling average seems like a better idea to me 🤷

Comment (dimitris-athanasiou) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/evaluation/common/AbstractAucRoc.java:107]: In all testing I've done the odd spikes were due to equal threshold points. I don't think I can provide a mathematical proof why we won't get problems for points with different threshold value but here is the intuition. As we scan for points, we only move in one direction. Interpolating in the other curve, we will also be moving in one direction. Thus I don't think we can ever get in trouble from points with different threshold.

**Review by benwtrent — APPROVED** [review]

**Comment by dimitris-athanasiou** [comment]
@elasticmachine update branch

**Pull request merged**